stdpopsim/cli.py

```diff
@@ -217,13 +218,60 @@ def write_output(ts, args):
         ts.dump(args.output)
 
 
-def write_citations(engine, model, contig):
+# The following are exceptions and mocks for
+# the testing of bibtex output.
+class InvalidCitation(Exception):
+    """Exception for a malformed or missing DOI."""
+    pass
+
+
+class ConnectionFailed(Exception):
+    """Exception for a failed bibtex retrieval."""
+    pass
+
+
+class mocked_get_response():
+    """
+    This exists solely to mock out the response from the
+    requests.get() call so that we can test the bibtex
+    retrieval."""
+    def __init__(self, ok, content, status_code):
+        self.ok = ok
+        self.content = content
+        self.status_code = status_code
+
+
+def get_bibtex(citation):
+    """Retrieve the bibtex of a citation from Crossref."""
+    try:
+        bib = get(citation.doi, headers={"Accept": "text/bibliography; style=bibtex"})
```

Comment:
Ah, I do see that now. Sorry about that, as you say this could definitely be easily done in `urllib` instead. I'll see if I can easily convert it over.

Reply:
No need to apologise Chris, this is great work!